File: C++/0128-longest-consecutive-sequence.cpp
// Tags: Array Amazon
// Time O(n)
// Space O(n)
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> numSet(nums.begin(), nums.end());
        int best = 0;
        for (auto x : nums) {
            if (!numSet.count(x - 1)) {
                int y = x + 1;
                while (numSet.count(y))
                    ++y;
                best = max(best, y - x);
            }
        }
        return best;
    }
};
/*
Then go through the numbers. 

If the number x is the start of a streak (i.e., x-1 is not in the set),
then test y = x+1, x+2, x+3, ... and stop at the first number y not in
the set. The length of the streak is then simply y-x and we update our
global best with that.

Since we check each streak only once, this is overall O(n)
*/



// Time O(n)
// Space O(n)
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        if (nums.empty())
            return 0;

        unordered_map<int,int> hash;
        int ans{1};
        for (const auto &n: nums) {
            if (!hash[n]) {
                hash[n] = 1;
                int leftbound {hash[n-1]}, rightbound {hash[n+1]};
                hash[n - leftbound] = hash [n + rightbound] = 1 + leftbound + rightbound;
                ans = max(ans, 1 + leftbound + rightbound);
            }
        }
        return ans;
    }
};

File: Python/0128-longest-consecutive-sequence.py
# Time O(n)
# Space O(n)


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        result, nums = 0, set(nums)

        for x in nums:
            if x - 1 not in nums:
                y = x + 1
                while y in nums:
                    y += 1
                result = max(result, y - x)
        return result


# Notes:
# Starting from each element x, we go as far as the count of consecutive
# elements can go.
# We only attempt to build sequences from numbers that
# are not already part of a longer sequence, ie only if
# x-1 is not in the set.
# O(n) because we skip if x-1 is present in the set and only iterative once.
# corner example : 6, 5, 4, 3, 2, 1
